### ISSUE TITLE
Clear expression and type equality definitions

### ISSUE THREAD

**Issue by Bastacyclop** [rise-lang/shine#77]
We need to properly discuss equality for expressions and types so that we have one or if necessary multiple clear definitions.
I believe that we need multiple definitions, depending on the context, but we should clarify them and discuss what should be the default one (`==`).

**Comment by Bastacyclop** [comment]
to solve: https://github.com/rise-lang/shine/pull/76#discussion_r515998779

**Comment by fedepiz** [comment]
As far as I can see, I can think of three type of equalities:

1. **Syntactic Equality**, which cares about differences such as variable names
2. **Rename-Equality**, which does not care about the above.
3. **Reference-Equality**, which cares about the memory location of the actual data structures used in the IR.

I suppose `3` in general is not what we care about, and we should avoid contaminating `1`and `2`with it by mistake if possible.

`2` is definitely necessary, and it's currently what the `==` operator for rise types and shine expressions and types is doing (as far as I know) (but it is NOT what ArithExpr == does as far as I know). The fact that `==` is rename equality is used in many places throughout the compiler.  This may cause issues when elements are put in `Map` and `Set` and other places that expect the property that `if x == y, then x.hashCode() == y.hashCode()` - which by our funky definition of `==` is not true

1` may be useful sometimes, comes out as the default `==` implementation of `case class`, and has the proper `hashCode` behaviour. So perhaps `1` should be `==`, and we would want to move `2` to some other operator - and then fix the inevitable deluge logic errors generated by changing the definition of `==`

Am I missing something, or is this picture correct and complete?

**Comment by Bastacyclop** [comment]
There are at least two other notions to consider:
- **"Open Equality"** how do we deal with free variables? currently differently named free variables are considered different unless you make the expression closed: https://github.com/rise-lang/shine/blob/e2e1039bdf4ae98d5af1feb0cfa87d928297ba0b/src/test/scala/rise/elevate/circularBuffering.scala#L110-L114
- **"Normal Form Equality"** where a normal form (e.g. BENF) is used as well: https://github.com/rise-lang/shine/blob/e2e1039bdf4ae98d5af1feb0cfa87d928297ba0b/src/test/scala/apps/separableConvolution2DRewrite.scala#L38-L43

**Comment by umazalakain** [comment]
Talking with Michel, we identified some of the equivalence relations at play.

On types, we have:
- (t0) syntactic equality (original `.equals`)
- (t1) alpha renaming (currently our overriden `.equals`)
- (t2) modulo unification, where type placeholders are equal to everything, and alpha renaming (the `=~=` relation)
- (t3) type erasure, all types are equal

On expressions, we have:
- (e0) syntactic equality (original `.equals`)
- (e1) alpha renaming (currently our override `.equals`, but without `=~=`)

Each of these equivalence relations needs of its corresponding hash function. I am wondering if it would make sense to parametrise the equality on expressions (and its corresponding hashing function) with an equality on types (and its corresponding hashing function). If you want to keep expression identifiers as keys in a map you probably want a combination of e0 with t2 (or maybe t3 if you are feeling adventurous). If you want to keep a set with more complicated expressions you probably want e1 together with t2 or t3.  After type inference t1 and t2 are presumably the same. Both t1 and e1 can be implemented by keeping a context of equivalent names instead of lifting into scala and performing the substitution.

**Comment by Bastacyclop** [comment]
@umazalakain sounds like a good summary, I would add, as discussed above:
- (t4) "open equality", where we want to allow differently named free variables to be equal on top of alpha renaming, so `n.dt1 == m.dt2` if we assume `n == m` and `dt1 == dt2`. I guess this is mostly useful for types and not so much for expressions, when rewriting openly typed terms.
- (e2) "normal form equality" where a normal form (e.g. BENF) is used as well. I think that this equality is fine to be implemented on the side (e.g. in Elevate) without a hash function.

**Comment by umazalakain** [comment]
If we implement alpha equivalence by keeping a list of identifiers considered equal that would allow us to implement "open equality" as an instance of alpha equivalence, right?

**Comment by Bastacyclop** [comment]
yes, it would require very little change, maybe just a flag like `unifyFreeIdentifiers` to allow putting unbound equivalences in the map.

**Comment by umazalakain** [comment]
Closed by #113
